src/backend/fs.rs

```diff
@@ -3,151 +3,127 @@
-    write_catalog(&file_hash, &segment_hashes)?;
+    write_catalog(&file_hash, segment_hashes)?;
 
     debug!("Finished write_file");
     Ok(file_hash)
 }
 
-#[allow(clippy::too_many_arguments)]
-pub fn write_segment(
-    sk: &[u8],
-    segment_path: PathBuf,
-    hash: &[u8; 32],
-    format: u8,
-    segment: &[u8],
-    chunk_index: usize,
-    encoded_len: u32,
-    padding_len: u32,
-) -> Result<()> {
-    let format = Format::try_from(format)?;
-    let header = Header::new(
-        sk,
-        hash,
-        format,
-        chunk_index as u8,
-        encoded_len,
-        padding_len,
-    )?;
-    let header_bytes = header.try_to_vec()?;
-    let file_name = header.file_name();
-    let path = segment_path.join(file_name);
-
-    trace!("Write segment at {}", path.to_string_lossy());
-    let mut file = OpenOptions::new()
-        .create_new(true)
-        .write(true)
-        .open(&path)?;
-
-    file.write_all(&header_bytes)?;
-    file.write_all(segment)?;
-    file.flush()?;
+pub fn write_segment(sk: &[u8], pk: &[u8], encoded: &Encoded) -> Result<BaoHash> {
+    let Encoded(encoded_bytes, bao_hash, encode_info) = encoded;
+    trace!("Encoded bytes len: {}", encoded_bytes.len());
 
-    Ok(())
+    let encoded_chunk_size = encode_info.bytes_verifiable as usize / SYS_CFG.volumes.len();
+    trace!("Encoded chunk size: {}", encoded_chunk_size);
+
+    encoded_bytes
+        .par_chunks_exact(encoded_chunk_size)
+        .enumerate()
+        .map(|(chunk_index, encoded_segment_chunk)| {
+            let volume = SYS_CFG
+                .volumes
+                .get(chunk_index)
+                .expect("Get one of eight volumes");
+
+            let format = Format::try_from(NODE_FORMAT)?;
+            let header = Header::new(
+                sk,
+                pk,
+                bao_hash.as_bytes(),
+                format,
+                chunk_index as u8,
+                encode_info.output_len,
+                encode_info.padding_len,
+            )?;
+            let header_bytes = header.try_to_vec()?;
+            let file_name = header.file_name();
+
+            let volume = SYS_CFG
+                .volumes
+                .get(chunk_index)
+                .expect("Get one of eight volumes");
+
+            let path = volume.path.join(SEGMENT_DIR).join(file_name);
+
+            trace!("Write segment at {}", path.to_string_lossy());
+            let mut file = OpenOptions::new()
+                .create_new(true)
+                .write(true)
+                .open(&path)?;
+
+            file.write_all(&header_bytes)?;
+            file.write_all(encoded_segment_chunk)?;
+            file.flush()?;
+
+            Ok(())
+        })
+        .collect::<Result<Vec<()>>>()?;
+
+    Ok(BaoHash(bao_hash.to_owned()))
 }
 
-pub fn write_catalog(file_hash: &Blake3Hash, segment_hashes: &[BaoHash]) -> Result<()> {
+pub fn write_catalog(file_hash: &Blake3Hash, segment_hashes: Vec<BaoHash>) -> Result<()> {
```

Comment:
&[BaoHash] becomes Vec<BaoHash> because of streaming?

Reply:
I can't remember why I made it an owned vec. I might change that in another update.